### PR TITLE
cut detection subsampler

### PR DESCRIPTION
**Pull request by danielmend**
lets us automatically detect cuts in video and subsample into multiple small clips using the ClippingSubsampler

### CODE REVIEW

**Comment by danielmend** [comment]
this doesnt work yet but once i put this in worker.py and test it out a bit ill ask for a proper review

File: video2dataset/subsamplers/cut_detection_subsampler.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+from scenedetect import detect, AdaptiveDetector
+
+
+class CutDetectionSubsampler:
+    def __init__(self, oom_clip_count, encode_formats, video_fps):
+        self.oom_clip_count = oom_clip_count
+        self.encode_formats = encode_formats
+        self.video_fps = video_fps
+
+        self.clip_subsampler = ClippingSubsampler(oom_clip_count, encode_formats)
```
Comment (iejMac) [inline video2dataset/subsamplers/cut_detection_subsampler.py:12]: subsamplers shouldn't use other subsamplers, you want to implement this in worker.py

**Comment by iejMac** [comment]
This is where the clipping subsampler is called - https://github.com/iejMac/video2dataset/blob/01c04aa9b2e2e43e3038a3c219a378068b22bc54/video2dataset/worker.py#L185

You want to do [something like with subtitles](https://github.com/iejMac/video2dataset/blob/01c04aa9b2e2e43e3038a3c219a378068b22bc54/video2dataset/worker.py#L179).  Basically the idea is you want to make a parameter which tells you if you want cut detection to determine clipping. If so you want to set the meta["clips"] in worker.py like in that if statement in linked. So I'm thinking you do something like:
```python3
if self.cut_split: # or idk maybe think of a better name for this var
    meta["clips"]= self.cut_detector(streams)
```

And like make sure the cut_detector returns the correct format so it works with the ClippingSubsampler. If you do this it should just work

File: video2dataset/main.py [hunk 2]
```diff
@@ -48,6 +48,7 @@ def video2dataset(
     tmp_dir: str = "/tmp",
     yt_metadata_args: dict = None,
     captions_are_subtitles: bool = False,
+    cut_mode: str = None,
```
Comment (iejMac) [inline video2dataset/main.py:51]: maybe cut_detection_mode, need to be expressive

**Comment by danielmend** [comment]
so there is a weird bug now where sometimes the clips array looks like [start, end] if theres only a single clip and sometimes it looks like [[start, end], [start, end], ...] and this is causing me issues when i switch between no cut detection and some cut detection. so im gonna debug that in detail tomorrow. also, @iejMac i discussed with andreas that we might want to detect cuts at various framerates -- this should be pretty simple i think, maybe we can just iterate over various framerates and write like {input}_{fps}.mp4 in the subsampler and then do cut detection on those lower-framerate videos in the tempdir, not sure. also andreas said it might be best if instead of cutting up the videos in v2ds we cut out the contiguous subsections in the dataloader step of whoever's using this data, that way we just have to keep track of metadata which allows for more flexibility.

File: tests/test_subsamplers.py [hunk 3]
```diff
@@ -139,3 +140,32 @@ def test_audio_rate_subsampler(sample_rate):
...
+    if cut_detection_mode == "longest":
+        assert len(cuts["cuts_original_fps"]) == 1
+        assert cuts["cuts_original_fps"] == [[0, 2096]]
+
+        if len(framerates) > 0:
+            assert cuts["cuts_15"] == [[0, 2096]]
+
+    if cut_detection_mode == "all":
+        assert len(cuts["cuts_original_fps"]) > 1
+        assert cuts["cuts_original_fps"][-1] == [3015, 3677]
```
Comment (iejMac) [inline tests/test_subsamplers.py:168]: just curious, how do you get these numbers? [3015, 3677]

Comment (danielmend) [reply tests/test_subsamplers.py:168]: i downloaded the test video and ran pyscenedetect manually on a google colab to get all these numbers

Comment (iejMac) [reply tests/test_subsamplers.py:168]: does it make sense though? I understand that it's good to have so we can confirm things in future PR's don't change this but is this result actually correct?

Comment (danielmend) [reply tests/test_subsamplers.py:168]: yea. look at the test video, its at 30fps so this cut is ~3000/30fps  = 100 seconds into the video . theres a cut (switching slides in a slideshow) at 1:40 so this checks out imo

File: video2dataset/subsamplers/cut_detection_subsampler.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    Detects cuts in input videos and returns contiguous segments in a video as metadata.
+
+    expects:
+    - cut_detection_mode to be either "longest" to pick the longest cut or "all" to pick all cuts
+    - framerates to be None (for original fps only) or a list of target framerates to detect cuts in
+    """
+
+    def __init__(self, cut_detection_mode="all", framerates=None):
+        if cut_detection_mode not in ["all", "longest"]:
```
Comment (iejMac) [inline video2dataset/subsamplers/cut_detection_subsampler.py:37]: remove this, let's make a default

File: video2dataset/main.py [hunk 5]
```diff
@@ -48,6 +48,9 @@ def video2dataset(
     tmp_dir: str = "/tmp",
     yt_metadata_args: dict = None,
     captions_are_subtitles: bool = False,
+    cut_detection_mode: str = None,
```
Comment (iejMac) [inline video2dataset/main.py:51]: set by default to longest

File: video2dataset/main.py [hunk 6]
```diff
@@ -48,6 +48,9 @@ def video2dataset(
...
+    cut_framerates: list = None,
```
Comment (iejMac) [inline video2dataset/main.py:52]: make sure this works if None is passed i.e. cut detection happens on original fps (if you already checked ignore this comment)

File: video2dataset/worker.py [hunk 7]
```diff
@@ -63,6 +74,11 @@ def __init__(
         self.data_reader = VideoDataReader(video_size, audio_rate, timeout, tmp_dir, yt_metadata_args, encode_formats)
 
         self.clipping_subsampler = ClippingSubsampler(oom_clip_count, encode_formats)
+        self.cut_detection_mode = cut_detection_mode
+        self.cut_framerates = cut_framerates
+        if cut_detection_mode is not None:
```
Comment (iejMac) [inline video2dataset/worker.py:79]: add a parameter "detect_cuts" which determines if we make this

Comment (iejMac) [reply video2dataset/worker.py:79]: also all main.py parameters need to be documented in README

File: video2dataset/main.py (hunk 6 above)
Comment (danielmend) [reply video2dataset/main.py:52]: im assuming you're referring to cut_framerates? in which case yea ive tested it a lot.

**Pull request merged**